reservation/actions.go

```diff
@@ -0,0 +1,184 @@
+package reservation
+
+import (
+	"context"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightninglabs/loop/fsm"
+	looprpc "github.com/lightninglabs/loop/swapserverrpc"
+)
+
+// InitReservationContext contains the request parameters for a reservation.
+type InitReservationContext struct {
+	ReservationID ID
+	ServerPubkey  *btcec.PublicKey
+	Value         btcutil.Amount
+	Expiry        uint32
+	HeightHint    uint32
+}
+
+type SweepRequest struct {
+	BestHeight uint32
+}
+
+// InitAction is the action that is executed when the reservation
+// state machine is initialized. It creates the reservation in the database
+// and dispatches the Payment to the server.
+func (r *FSM) InitAction(
+	eventCtx fsm.EventContext) fsm.EventType {
+
+	// Check if the context is of the correct type.
+	reservationRequest, ok := eventCtx.(*InitReservationContext)
+	if !ok {
+		return r.HandleError(fsm.ErrInvalidContextType)
+	}
+
+	keyRes, err := r.cfg.Wallet.DeriveNextKey(r.ctx, KeyFamily)
+	if err != nil {
+		return r.HandleError(err)
+	}
+
+	// Send the client reservation details to the server.
+	log.Debugf("dispatching reservation to server: %x",
+		reservationRequest.ReservationID)
+
+	request := &looprpc.ServerOpenReservationRequest{
+		ReservationId: reservationRequest.ReservationID[:],
+		ClientKey:     keyRes.PubKey.SerializeCompressed(),
+	}
+
+	_, err = r.cfg.ReservationClient.OpenReservation(
+		context.Background(), request,
+	)
+	if err != nil {
+		return r.HandleError(err)
+	}
+
+	reservation := &Reservation{
+		ID:               reservationRequest.ReservationID,
+		Value:            reservationRequest.Value,
+		ClientPubkey:     keyRes.PubKey,
+		ServerPubkey:     reservationRequest.ServerPubkey,
+		KeyLocator:       keyRes.KeyLocator,
+		Expiry:           reservationRequest.Expiry,
+		InitiationHeight: int32(reservationRequest.HeightHint),
+	}
+
+	r.reservation = reservation
```

Comment:
Can it cause any issues that we set `r.reservation` before ensuring that we were able to store it?

Reply:
Sorry my bad, I thought that `CreateReservation` has access to the `r.reservation`, so we could've omitted `r.ctx, reservation` in the method sig.